### PR TITLE
Updates

### PR DESCRIPTION
**Pull request by PhilipVinc**

### CODE REVIEW

File: mpi4jax/_src/collective_ops/barrier.py [hunk 1]
```diff
@@ -73,16 +73,14 @@ def mpi_barrier_xla_encode_cpu(ctx, token, comm):
     )
 
     # JAX insists on outputs being iterable
```
Comment (dionhaefner) [inline mpi4jax/_src/collective_ops/barrier.py:75]: Outdated comment

File: mpi4jax/_src/collective_ops/barrier.py [hunk 2]
```diff
@@ -98,17 +96,15 @@ def mpi_barrier_xla_encode_gpu(ctx, token, comm):
     descriptor = build_barrier_descriptor(to_mpi_handle(comm))
 
     # JAX insists on outputs being iterable
```
Comment (dionhaefner) [inline mpi4jax/_src/collective_ops/barrier.py:98]: Dito

File: mpi4jax/_src/collective_ops/send.py [hunk 3]
```diff
@@ -91,16 +91,14 @@ def mpi_send_xla_encode_cpu(ctx, x, token, dest, tag, comm):
     )
 
     # JAX insists on outputs being iterable
```
Comment (dionhaefner) [inline mpi4jax/_src/collective_ops/send.py:93]: Dito

File: mpi4jax/_src/collective_ops/send.py [hunk 4]
```diff
@@ -135,17 +133,15 @@ def mpi_send_xla_encode_gpu(ctx, x, token, dest, tag, comm):
...
```
Comment (dionhaefner) [inline mpi4jax/_src/collective_ops/send.py:135]: Dito

File: mpi4jax/_src/jax_compat.py [hunk 5]
```diff
@@ -59,8 +76,34 @@ def check_jax_version():
...
 
 
+# TODO: remove this code once we only support jax > 0.4.14
+if versiontuple(jax.__version__) >= (0, 4, 14):
+    from jax.core import ShapedArray  # noqa: F401
+else:
+    from jax.abstract_arrays import ShapedArray  # noqa: F401
+
+
+# TODO: remove this code once we only support jax > 0.4.14
```
Comment (dionhaefner) [inline mpi4jax/_src/jax_compat.py:86]:
```suggestion
# TODO: remove this code once we only support jax > 0.4.16
```

File: mpi4jax/_src/jax_compat.py [hunk 6]
```diff
@@ -59,8 +76,34 @@ def check_jax_version():
...
+            control_flow_allowed_effects,
+            custom_derivatives_allowed_effects,
+        )
+
+        effect = EffectType()
+        mlir.lowerable_effects.add_type(EffectType)
+        control_flow_allowed_effects.add_type(EffectType)
+        # Effects must be added to the allow_effects list in order to work within
+        # custom_vjp. See google/jax#11916
+        custom_derivatives_allowed_effects.add_type(EffectType)
```
Comment (dionhaefner) [inline mpi4jax/_src/jax_compat.py:101]: I see some more effects here:

https://github.com/google/jax/blob/7bc01d9472fbe872e14db59eed81a26b0e8c078b/jax/_src/debugging.py#L68-L76

Namely, `effects.ordered_effects`, `effects.lowerable_effects` (is this an alias for `mlir.lowerable_effects`?), `effects.remat_allowed_effects`. Especially `ordered_effects` sounds relevant for us?!

File: mpi4jax/_src/jax_compat.py [hunk 7]
```diff
@@ -59,8 +76,34 @@ def check_jax_version():
...
+        return effect
+
+
 # TODO: remove this code once we only support jax > 0.4.5
-if versiontuple(jax.__version__) >= (0, 4, 5):
+elif versiontuple(jax.__version__) >= (0, 4, 5):
```
Comment (dionhaefner) [inline mpi4jax/_src/jax_compat.py:106]: (We should do a deprecation cycle after merging this)

File: mpi4jax/_src/jax_compat.py (hunk 5 above)
Comment (PhilipVinc) [reply mpi4jax/_src/jax_compat.py:86]: I really meant `> 0.4.14` : `0.4.15` has been yanked so it is starting with 0.4.16... I'll just change this to `>0 0.4.16` which is more clear.

File: mpi4jax/_src/jax_compat.py (hunk 6 above)
Comment (PhilipVinc) [reply mpi4jax/_src/jax_compat.py:101]: I simply adapted the code following what was moved around. I suspect we might tag this as being allowed under remat, but would rather add a test later on than quickly add it here and risk getting segfaults.
In any case, remat is not supported even now.

Comment (dionhaefner) [reply mpi4jax/_src/jax_compat.py:101]: The module docstring here seems to imply that we can eliminate tokens if we use `ordered_effects`: https://github.com/google/jax/blob/main/jax/_src/effects.py

Big if true.

Comment (dionhaefner) [reply mpi4jax/_src/jax_compat.py:101]: Also, [`mlir.lowerable_effects` *is* an alias for `effects.lowerable_effects`](https://github.com/google/jax/blob/main/jax/_src/interpreters/mlir.py#L64), so we should be using that instead.

**Comment by PhilipVinc** [comment]
F tf

Il mer 20 set 2023, 11:20 AM Dion Häfner ***@***.***> ha
scritto:

File: mpi4jax/_src/jax_compat.py (hunk 6 above)
Comment (PhilipVinc) [reply mpi4jax/_src/jax_compat.py:101]: _LiesCalcAndCompTold.p_t_df_____b.com/mpi4jax/mpi4jax/files/12671790/LiesCalcAndCompTold.pdf)
e
[]()[]()_

**Pull request merged**